### PR TITLE
telemetry: init internal APIs and events cache store layer

### PR DESCRIPTION
**Pull request by bobheadxi**
This PR introduces:

1. a new backend telemetry API, with `internal/telemetry.EventRecorder`, for backend services to generate their own events, based on [the proposal](https://docs.google.com/document/d/14WBt80sbmVm73B-R1Srs5cSunZo2IhDMtKiyAgkYujU/edit#bookmark=id.wy4lr5f3lxyk)
   - this is similar to the clientside SDK: https://github.com/sourcegraph/telemetry
3. a caching layer in the database, `internal/database.TelemetryEventsExportQueueStore`, that stores raw Protobuf messages so that a worker can be implemented on top to pull events out and export them in batches
   - planned strategy is to mark exported entries as exported, and then periodically prune exported events from the table after N day(s)
4. an `internal/telemetry/teestore.Store` which tees events into the existing `event_logs` table (which we are no longer considering revamping due to extensive existing integrations with it) as well as the new `TelemetryEventsExportQueueStore`
   - adding things to this store is behind an off-by-default feature flag `telemetry-export` - when the flag is disabled, adding to the store is a no-op
5. adapters to send events from `telemetry.EventRecorder` and the GraphQL mutation added in https://github.com/sourcegraph/sourcegraph/pull/56297 to the `teestore.Store`

Actually exporting things, and the destination service itself, will be built in https://github.com/sourcegraph/sourcegraph/pull/56699
Stacked on https://github.com/sourcegraph/sourcegraph/pull/56519

Closes https://github.com/sourcegraph/sourcegraph/issues/56283
Closes https://github.com/sourcegraph/sourcegraph/issues/56285
Closes https://github.com/sourcegraph/sourcegraph/issues/56286

## Test plan

Unit and integration tests, and manual test plan in https://github.com/sourcegraph/sourcegraph/pull/56699

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #56297** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/56297" target="_blank">Graphite</a> 
    * **PR #56519** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/56519" target="_blank">Graphite</a> 
      * **PR #56520** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/56520" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/56520?utm_source=stack-comment).

File: cmd/frontend/internal/telemetry/resolvers/telemetrygateway.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		// Configure parameters
+		event.Parameters = &telemetrygatewayv1.EventParameters{
+			Version: e.Parameters.Version,
+			Metadata: func() map[string]int64 {
+				if e.Parameters.Metadata == nil || len(*e.Parameters.Metadata) == 0 {
+					return nil
+				}
+				metadata := make(map[string]int64, len(*e.Parameters.Metadata))
+				for _, kv := range *e.Parameters.Metadata {
+					metadata[kv.Key] = int64(kv.Value)
```
Comment (rafax) [inline cmd/frontend/internal/telemetry/resolvers/telemetrygateway.go:55]: Any reason why we use int64 in metadata, but int32 in kv.Value -> should we change on of those to make types match?

**Review by rafax — APPROVED** [review]
LGTM but really more of a stamp (for a 38 file) PR :)

File: cmd/frontend/internal/telemetry/resolvers/telemetrygateway.go (hunk 1 above)
Comment (bobheadxi) [reply cmd/frontend/internal/telemetry/resolvers/telemetrygateway.go:55]: For GraphQL/JSON to support int64, [we need to use `BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt), which is a little annoying - I'm unsure if we have a use case for recording massive numbers clientside 🤔

Comment (bobheadxi) [reply cmd/frontend/internal/telemetry/resolvers/telemetrygateway.go:55]: Checked in with @nathan-downs , he concurs that we're unlikely to need really big numbers clientside

File: migrations/frontend/1694806099_add_telemetryeventsexportqueuestore/up.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE IF NOT EXISTS  telemetry_events_export_queue (
+  id TEXT PRIMARY KEY,
+  timestamp TIMESTAMPTZ NOT NULL,
+  payload_pb BYTEA NOT NULL,
+  exported_at TIMESTAMPTZ DEFAULT NULL
+);
+
+CREATE INDEX IF NOT EXISTS telemetry_events_export_queue_id_idx ON telemetry_events_export_queue (id);
```
Comment (unknwon) [inline migrations/frontend/1694806099_add_telemetryeventsexportqueuestore/up.sql:8]: CleanShot 2023-09-19 at 14 42 58@2x
Wouldn't primary key comes with the same type of index by itself?

**Review by unknwon — APPROVED** [review]

File: migrations/frontend/1694806099_add_telemetryeventsexportqueuestore/up.sql (hunk 2 above)
Comment (bobheadxi) [reply migrations/frontend/1694806099_add_telemetryeventsexportqueuestore/up.sql:8]: Good point haha, I looked into this and they are indeed indexed in the same way automatically. Removing it!

**Review by nathan-downs — APPROVED** [review]

**Pull request merged**